File: example/scalalib/web/5-webapp-scalajs-shared/src/WebApp.scala
package webapp
import scalatags.Text.all._
import scalatags.Text.tags2
import shared.{Shared, Todo}

object WebApp extends cask.MainRoutes {
  override def port = 8083
  var todos = Seq(
    Todo(true, "Get started with Cask"),
    Todo(false, "Profit!")
  )

  @cask.post("/list/:state")
  def list(state: String) = upickle.default.write(todos)

  @cask.post("/add/:state")
  def add(state: String, request: cask.Request) = {
    todos = Seq(Todo(false, request.text())) ++ todos
    upickle.default.write(todos)
  }

  @cask.post("/delete/:state/:index")
  def delete(state: String, index: Int) = {
    todos = todos.patch(index, Nil, 1)
    upickle.default.write(todos)
  }

  @cask.post("/toggle/:state/:index")
  def toggle(state: String, index: Int) = {
    todos = todos.updated(index, todos(index).copy(checked = !todos(index).checked))
    upickle.default.write(todos)
  }

  @cask.post("/clear-completed/:state")
  def clearCompleted(state: String) = {
    todos = todos.filter(!_.checked)
    upickle.default.write(todos)
  }

  @cask.post("/toggle-all/:state")
  def toggleAll(state: String) = {
    val next = todos.filter(_.checked).size != 0
    todos = todos.map(_.copy(checked = next))
    upickle.default.write(todos)
  }

  @cask.get("/")
  def index() = {
    doctype("html")(
      html(lang := "en")(
        head(
          meta(charset := "utf-8"),
          meta(name := "viewport", content := "width=device-width, initial-scale=1"),
          tags2.title("Template • TodoMVC"),
          link(rel := "stylesheet", href := "/static/index.css")
        ),
        body(
          tags2.section(cls := "todoapp")(Shared.renderBody(todos, "all")),
          footer(cls := "info")(
            p("Double-click to edit a todo"),
            p("Created by ")(
              a(href := "http://todomvc.com")("Li Haoyi")
            ),
            p("Part of ")(
              a(href := "http://todomvc.com")("TodoMVC")
            )
          ),
          script(src := "/static/main.js")
        )
      )
    )
  }

  @cask.staticResources("/static")
  def static() = "webapp"

  initialize()
}

File: example/kotlinlib/web/6-webapp-kotlinjs-shared/src/WebApp.kt
package webapp

import io.ktor.serialization.kotlinx.json.json
import io.ktor.server.application.Application
import io.ktor.server.application.call
import io.ktor.server.application.install
import io.ktor.server.engine.embeddedServer
import io.ktor.server.html.respondHtml
import io.ktor.server.http.content.staticResources
import io.ktor.server.netty.Netty
import io.ktor.server.plugins.contentnegotiation.ContentNegotiation
import io.ktor.server.request.receiveText
import io.ktor.server.response.respond
import io.ktor.server.routing.get
import io.ktor.server.routing.post
import io.ktor.server.routing.routing
import io.ktor.server.util.getOrFail
import kotlinx.html.HTML
import kotlinx.html.a
import kotlinx.html.body
import kotlinx.html.footer
import kotlinx.html.head
import kotlinx.html.link
import kotlinx.html.meta
import kotlinx.html.p
import kotlinx.html.script
import kotlinx.html.section
import kotlinx.html.title
import shared.Todo
import shared.renderBody

object WebApp {
    private val todos =
        mutableListOf(
            Todo(true, "Get started with Cask"),
            Todo(false, "Profit!"),
        )

    fun add(
        state: String,
        text: String,
    ) {
        todos.add(Todo(false, text))
    }

    fun delete(
        state: String,
        index: Int,
    ) {
        todos.removeAt(index)
    }

    fun toggle(
        state: String,
        index: Int,
    ) {
        todos[index] =
            todos[index].let {
                it.copy(checked = !it.checked)
            }
    }

    fun clearCompleted(state: String) {
        todos.removeAll { it.checked }
    }

    fun toggleAll(state: String) {
        val next = todos.any { !it.checked }
        for (item in todos.withIndex()) {
            todos[item.index] = item.value.copy(checked = next)
        }
    }

    private fun HTML.renderIndex() {
        head {
            meta(charset = "utf-8")
            meta(name = "viewport", content = "width=device-width, initial-scale=1")
            title("Template • TodoMVC")
            link(rel = "stylesheet", href = "/static/index.css")
        }
        body {
            section(classes = "todoapp") {
                renderBody(todos, "all")
            }
            footer(classes = "info") {
                p { +"Double-click to edit a todo" }
                p {
                    +"Created by "
                    a(href = "http://todomvc.com") { +"Li Haoyi" }
                }
                p {
                    +"Part of "
                    a(href = "http://todomvc.com") { +"TodoMVC" }
                }
            }
            script(src = "/static/client.js", block = {})
        }
    }

    fun configureRoutes(app: Application) {
        with(app) {
            routing {
                get("/") {
                    call.respondHtml {
                        renderIndex()
                    }
                }
                post("/toggle-all/{state}") {
                    toggleAll(call.parameters.getOrFail("state"))
                    call.respond(todos)
                }
                post("/clear-completed/{state}") {
                    clearCompleted(call.parameters.getOrFail("state"))
                    call.respond(todos)
                }
                post("/toggle/{state}/{index}") {
                    call.parameters.run {
                        toggle(getOrFail("state"), getOrFail<Int>("index"))
                        call.respond(todos)
                    }
                }
                post("/delete/{state}/{index}") {
                    call.parameters.run {
                        delete(getOrFail("state"), getOrFail<Int>("index"))
                        call.respond(todos)
                    }
                }
                post("/add/{state}") {
                    val requestText = call.receiveText()
                    add(call.parameters.getOrFail("state"), requestText)
                    call.respond(todos)
                }
                post("/list/{state}") {
                    call.respond(todos)
                }
                staticResources("/static", "webapp")
            }
        }
    }

    @JvmStatic
    fun main(args: Array<String>) {
        embeddedServer(Netty, port = 8093, host = "0.0.0.0") {
            install(ContentNegotiation) {
                json()
            }
            configureRoutes(this)
        }.start(wait = true)
    }
}
